Support for trail filters

AndreasAZiegler opened the pull request:
This PR adds support for the trail filters [described here](https://docs.prophesee.ai/stable/hw/manuals/esp.html#event-trail-filter-stc-trail). Successfully tested on an EVKv4 from Prophesse.

Note: So far I only implemented it for ROS1.

berndpfrommer commented:
Hi Andreas,
Thanks for the PR!
First let me apologize for leaving an earlier PR sitting that I just merged. I had simply forgotten to merge it in, it was ready to go. Apparently it had some conflicts, but it should be easy to resolve.
Second, I have some stylistic changes that I'd like to see before we merge this. Sorry for being peculiar, but I think the code can be written shorter with minimum performance penalty. A smaller code base is easier to maintain, and has less potential for bugs. Please see feedback in the comments.
Just to clarify: this is a valid and good PR, and I want it going into the driver, it just needs some polishing.

berndpfrommer commented on include/metavision_driver/metavision_wrapper.h:
```diff
@@ -60,6 +60,15 @@ class MetavisionWrapper
 
+  struct TrailFilter
+  {
+    enum Type { TRAIL, STC_CUT_TRAIL, STC_KEEP_TRAIL };
```
This enum is not needed. Keep type as a string, and use std::map to map to the metavision SDK enums at the very last moment.

berndpfrommer commented on include/metavision_driver/metavision_wrapper.h:
```diff
@@ -112,6 +121,7 @@ class MetavisionWrapper
     ercRate_ = rate;
   }
   void setMIPIFramePeriod(int usec) { mipiFramePeriod_ = usec; }
+  void setTrailFilter(const TrailFilter::Type type, const uint32_t threshold, const bool state);
```
change type to string

berndpfrommer commented on src/driver_ros2.cpp:
```diff
@@ -317,6 +317,24 @@ void DriverROS2::configureWrapper(const std::string & name)
+      filter_type = metavision_driver::MetavisionWrapper::TrailFilter::Type::STC_CUT_TRAIL;
+    } else if (trailFilterType == "stc_keep_cut") {
+      filter_type = metavision_driver::MetavisionWrapper::TrailFilter::Type::STC_KEEP_TRAIL;
+    }
```
There is no final "else" clause, so filter_type will be left undefined.
I suggest removing all of this code, and passing trailFilterType straight through to the metavision wrapper

AndreasAZiegler replied:
But in this case, the ROS2 wrapper would need to be exposed to some metavision headers which it is not yet, afaik. This was the motivation between this "intermediate" type defined in the wrapper. Or is this related to a previous comment of using strings and mapping it at the end?

berndpfrommer commented on src/metavision_wrapper.cpp:
```diff
@@ -119,6 +120,15 @@ int MetavisionWrapper::setBias(const std::string & name, int val)
 }
 
+void MetavisionWrapper::setTrailFilter(
+  const TrailFilter::Type type, const uint32_t threshold, const bool state)
```
change first argument type to string

berndpfrommer commented on src/metavision_wrapper.cpp:
```diff
@@ -428,8 +438,38 @@ void MetavisionWrapper::setDecodingEvents(bool decodeEvents)
+    filter_type = Metavision::I_EventTrailFilterModule::Type::STC_CUT_TRAIL;
+  } else if (trailFilter_.type == TrailFilter::Type::STC_KEEP_TRAIL) {
+    filter_type = Metavision::I_EventTrailFilterModule::Type::STC_KEEP_TRAIL;
+  }
```
Use a const static std::map<std::string, Metavision::I_EventTrailFilterModule::Type> map that you initialize with {}. This will make the code more readable. Then it = map.find(type) and throw exception if it == map.end().
Note again here you are not handling the case where all if()'s evaluate to false (but then you are using an enum, one can argue it's not necessary). Either way, this code should go in favor of a map.

berndpfrommer reviewed (changes requested):
Thanks again for the PR, please see review comments.

berndpfrommer commented:
Oh, and you wrote you only did ROS1, but the PR has only ROS2. That is fine. ROS1 is basically end-of-life.

AndreasAZiegler commented:
@berndpfrommer Thanks a lot for the valuable feedback, I'm always happy to learn and improve. It has been a while for me since the last PR with a "senior" developer as reviewer.
I believe I addressed all of your comments. Please let me know if I missed some. I hope the test pass as well.

berndpfrommer commented on include/metavision_driver/metavision_wrapper.h:
```diff
@@ -16,6 +16,7 @@
 #ifndef METAVISION_DRIVER__METAVISION_WRAPPER_H_
 #define METAVISION_DRIVER__METAVISION_WRAPPER_H_
 
+#include <metavision/hal/facilities/i_event_trail_filter_module.h>
```
Can this be moved to ``metavision_wrapper.cpp``?

berndpfrommer commented on src/metavision_wrapper.cpp:
```diff
@@ -428,8 +442,44 @@ void MetavisionWrapper::setDecodingEvents(bool decodeEvents)
+
+  Metavision::I_EventTrailFilterModule::Type filter_type;
+
+  /*
```
Please remove this comment.

berndpfrommer commented on src/metavision_wrapper.cpp:
```diff
@@ -428,8 +442,44 @@ void MetavisionWrapper::setDecodingEvents(bool decodeEvents)
+  */
+  const auto it = trailFilterMap.find(trailFilter_.type);
+  if (it == trailFilterMap.end()) {
+    LOG_WARN_NAMED("unknown trail filter type " << trailFilter_.type);
```
call BOMB_OUT() here (logs and throws exception). As it stands, this is a bug because if you don't abort here, you will be dereferencing an invalid iterator.

berndpfrommer commented:
Tests fail due to formatting errors. You can check in your workspace before committing by:
```
colcon test --packages-select metavision_driver
colcon test-result --verbose
```
Also please see additional comments: bomb out if invalid filter type is specified. You want to get the user's attention when they are feeding in nonsense parameters.
Another few nitpicks as well, see new comments.

AndreasAZiegler commented:
Is there something I can help to get the test passing? Looks like there some issues in `logging.h`?
